src/target/puya.c

```diff
@@ -0,0 +1,171 @@
+		return false;
+	}
+
+	target_add_ram32(target, RAM_START, ram_size);
+	target_flash_s *flash = calloc(1, sizeof(*flash));
+	if (!flash) { /* calloc failed: heap exhaustion */
+		DEBUG_ERROR("calloc: failed in %s\n", __func__);
+		return false;
+	}
+	flash->start = FLASH_START;
+	flash->length = flash_size;
+	flash->blocksize = FLASH_PAGE_SIZE;
+	flash->writesize = FLASH_PAGE_SIZE;
+	flash->erase = puya_flash_erase;
+	flash->write = puya_flash_write;
+	flash->prepare = puya_flash_prepare;
+	flash->done = puya_flash_done;
+	flash->erased = 0xffU;
+	target_add_flash(target, flash);
+
+	return false;
+}
+
+static bool puya_flash_prepare(target_flash_s *flash)
+{
+	target_mem32_write32(flash->t, FLASH_KEYR, FLASH_KEYR_KEY1);
+	target_mem32_write32(flash->t, FLASH_KEYR, FLASH_KEYR_KEY2);
+
+	int hsi_fs = (target_mem32_read32(flash->t, RCC_ICSCR) >> 13) & 7;
+	if (hsi_fs > 4)
+		hsi_fs = 0;
+	DEBUG_INFO("HSI frequencey selection is %d\n", hsi_fs);
+
+	const uint32_t eppara0 = target_mem32_read32(flash->t, FLASH_TIMING_CAL_BASE + hsi_fs * 20 + 0);
+	const uint32_t eppara1 = target_mem32_read32(flash->t, FLASH_TIMING_CAL_BASE + hsi_fs * 20 + 4);
+	const uint32_t eppara2 = target_mem32_read32(flash->t, FLASH_TIMING_CAL_BASE + hsi_fs * 20 + 8);
+	const uint32_t eppara3 = target_mem32_read32(flash->t, FLASH_TIMING_CAL_BASE + hsi_fs * 20 + 12);
+	const uint32_t eppara4 = target_mem32_read32(flash->t, FLASH_TIMING_CAL_BASE + hsi_fs * 20 + 16);
+	DEBUG_INFO("PY32 flash timing cal 0: %08" PRIx32 "\n", eppara0);
+	DEBUG_INFO("PY32 flash timing cal 1: %08" PRIx32 "\n", eppara1);
+	DEBUG_INFO("PY32 flash timing cal 2: %08" PRIx32 "\n", eppara2);
+	DEBUG_INFO("PY32 flash timing cal 3: %08" PRIx32 "\n", eppara3);
+	DEBUG_INFO("PY32 flash timing cal 4: %08" PRIx32 "\n", eppara4);
+
+	target_mem32_write32(flash->t, FLASH_TS0, eppara0 & 0xff);
+	target_mem32_write32(flash->t, FLASH_TS1, (eppara0 >> 16) & 0x1ff);
+	target_mem32_write32(flash->t, FLASH_TS3, (eppara0 >> 8) & 0xff);
+	target_mem32_write32(flash->t, FLASH_TS2P, eppara1 & 0xff);
+	target_mem32_write32(flash->t, FLASH_TPS3, (eppara1 >> 16) & 0x7ff);
+	target_mem32_write32(flash->t, FLASH_PERTPE, eppara2 & 0x1ffff);
+	target_mem32_write32(flash->t, FLASH_SMERTPE, eppara3 & 0x1ffff);
+	target_mem32_write32(flash->t, FLASH_PRGTPE, eppara4 & 0xffff);
+	target_mem32_write32(flash->t, FLASH_PRETPE, (eppara4 >> 16) & 0x3fff);
+
+	return true;
+}
+
+static bool puya_flash_done(target_flash_s *flash)
+{
+	target_mem32_write32(flash->t, FLASH_CR, FLASH_CR_LOCK);
+	return true;
+}
+
+static bool puya_wait_flash(target_s *const target, platform_timeout_s *const timeout)
+{
+	while (target_mem32_read32(target, FLASH_SR) & FLASH_SR_BSY) {
+		if (target_check_error(target))
+			return false;
+		if (timeout)
+			target_print_progress(timeout);
+	}
+	return true;
+}
+
+static bool puya_flash_erase(target_flash_s *flash, target_addr_t addr, size_t len)
+{
+	(void)len;
+	DEBUG_INFO("PY32 flash erase @ %08" PRIx32 "\n", addr);
```

Comment:
Please drop this and the DEBUG_INFO in write - suitable debugging is already in the target Flash layer, so this duplicates that.